proposals/NNNN-parameter-packs.md

```diff
@@ -0,0 +1,754 @@
+The downsides to postfix `*` include:
+
+* `*` is extremely subtle
+* `*` evokes pointer types / a dereferencing operator to programmers familiar with other languages including C/C++, Go, Rust, etc.
+* Choosing another operator does not alleviate the ambiguities in expressions, because values could also have a postfix `*` operator or any other operator symbol, leading to the same ambiguity.
+
+The downsides to introducing keywords are:
+
+* Though the keywords are more verbose than an operator, using the `expand` keyword in expression context is still fairly subtle because it looks just like a function call rather than a built in expansion operation.
+* Introducing a new keyword in expression context would break existing code that uses that keyword name, e.g. as the name of a function
+
+#### Magic builtin `map` method
+
+The prevlence of `map` and `zip` in Swift makes this syntax an attractive option for variadic generics:
+
+```swift
+func wrap<each T>(_ values: repeat each T) -> (repeat Wrapped<each T>) {
+  return values.map { Wrapped($0) }
+}
+```
+
+The downsides of a magic `map` method are:
+
+* `.map` isn't only used for mapping elements to a new element, it's also used for direct forwarding, which is very different to the way `.map` is used today. An old design exploration for variadic generics used a map-style builtin, but allowed exact forwarding to omit the `.map { $0 }`. Privileging exact forwarding would be pretty frustrating, because you would need to add `.map { $0 }` as soon as you want to append other elements to either side of the pack expansion, and it wouldn't work for other values that you might want to turn into packs such as tuples or arrays.
+* There are two very different models for working with packs; the same conceptual expansion has very different spellings at the type and value level, `repeat Wrapped<each T>` vs `values.map { Wrapped($0) }`.
+* Magic map can only be applied to one pack at a time, leaving no clear way to zip packs without adding other builtins. A `zip` builtin would also be misleading, because expanding two packs in parallel does not need to iterate over the packs twice, but using `zip(t, u).map { ... }` looks that way.
+* The closure-like syntax is misleading because it’s not a normal closure that you can write in the language. This operation is also very complex over packs with any structure, including concrete types, because the compiler either needs to infer a common generic signature for the closure that works for all elements, or it needs to separately type check the closure once for each element type.
+* `map` would still need to be resolved via overload resolution amongst the existing overloads, so this approach doesn't help much with the type checking issues that `...` has.
+
+## Future directions
+
+### Variadic generic types
+
+This proposal only supports type parameter packs on functions. A complementary proposal will describe type parameter packs on generic structs, enums and classes.
+
```

Comment:
Since it seems like all the stuff about `for...in` iterating over a value pack is no longer part of this proposal, should it be mentioned here as a future direction?

Relatedly, indexing into a pack was mentioned during the first pitch.

Reply:
Yeah, we subsetted pack iteration out of this first proposal. I added pack iteration and pack element projection (both dynamic indexing and statically typed indexing) to the future directions.